daemon/cmd/daemon.go

```diff
@@ -1058,13 +1057,21 @@ func newDaemon(ctx context.Context, cleaner *daemonCleanup, params *daemonParams
 	// router IPs from the K8s resources if we weren't able to restore them
 	// from the fs. We must do this after IPAM because we must wait until the
 	// K8s resources have been synced. Part 2/2 of restoration.
+	var restoredRouterIPv4, restoredRouterIPv6 net.IP
+	if option.Config.EnableIPv4 {
+		restoredRouterIPv4 = d.restoreCiliumHostIPs(false, router4FromK8s)
+	}
+	if option.Config.EnableIPv6 {
+		restoredRouterIPv6 = d.restoreCiliumHostIPs(true, router6FromK8s)
+	}
+
 	gcHostIPsFn := func(ctx context.Context, retries int) (done bool, err error) {
 		var errs error
-		if option.Config.EnableIPv4 {
-			errs = errors.Join(errs, d.restoreCiliumHostIPs(false, router4FromK8s))
+		if restoredRouterIPv4 != nil {
```

Comment:
In the previous case:

```
	restoredIP := node.RestoreHostIPs(ipv6, fromK8s, fromFS, cidrs)
	return removeOldRouterState(ipv6, restoredIP)
```

I believe removeOldRouterState would always be attempted for the respective ip family, even if the restoredIP was nil.
Should this also always attempt remove old router state even if there is no restored IP, looking at the removeOldRouterState code that it was designed to handle this [case](https://github.com/cilium/cilium/blob/main/daemon/cmd/daemon.go#L375)

Reply:
The code you highlighted is from an intermediate commit - in the final version `removeOldRouterState` should be always called with the real (not the restored) router IP:
https://github.com/cilium/cilium/blob/26308e2550cc533fc14fb800047a3d47ee6f0945/daemon/cmd/ipam.go#L486-L487

The final version should (at least in theory) always clean up the interface, unless the IP family is disabled. That latter condition however was also present in the old version of the code ([see here, where restoreCiliumHostIPs later calls into removeOldRouterState](https://github.com/cilium/cilium/blob/5ec730f5e22a7af45f16d7cc593e755ed8265dd2/daemon/cmd/daemon.go#L1042-L1047). 

I can try to fix up the intermediate commit, but it shouldn't change the final outcome.